pkg/devfile/components.go

```diff
@@ -50,3 +50,56 @@ func GetKubernetesComponentsToPush(devfileObj parser.DevfileObj, allowApply bool
 
 	return k8sComponents, err
 }
+
+// GetApplyComponentsToPush returns devfile K8s components associated with the given commandName and commandGroupKind;
+// these resources will always be referenced with an apply command, hence the function name
+func GetApplyComponentsToPush(devfileObj parser.DevfileObj, commandGroupKind devfilev1.CommandGroupKind, commandName string, wantK8sComponent, wantImageComponent bool) (k8sComponentNames, imageComponentNames []string, _ error) {
+	commands, err := devfileObj.Data.GetCommands(parsercommon.DevfileOptions{
+		CommandOptions: parsercommon.CommandOptions{
+			CommandGroupKind: commandGroupKind,
+			CommandType:      devfilev1.CompositeCommandType,
+		},
+		FilterByName: commandName,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	if len(commands) == 0 {
+		return nil, nil, nil
+	}
+	// we assume there will be only one command that matches the command ID
+	command := commands[0]
+	for _, subCommand := range command.Composite.Commands {
+		cmd, err := devfileObj.Data.GetCommands(parsercommon.DevfileOptions{
+			FilterByName:   subCommand,
+			CommandOptions: parsercommon.CommandOptions{CommandType: devfilev1.ApplyCommandType},
+		})
+		if err != nil {
+			return nil, nil, err
+		}
+		if len(cmd) == 0 {
+			continue
+		}
+		// we assume there will be only one command that matches the subCommand ID
+		if applyCmd := cmd[0]; applyCmd.Apply != nil {
```

Comment:
This assumes that the `apply` component will be a direct child of the composite command. But we might have several levels, I think, e.g., a composite command containing another composite command in turn containing another composite command referencing an `apply` Kubernetes component. And in such cases, no warning is displayed.

What do you think about displaying the warning messages in the [Podman command handler](https://github.com/redhat-developer/odo/blob/main/pkg/dev/podmandev/command.go) instead, while the handler is walking through each component when running the command? You would still need to handle the case of K8s components not referenced by any command.

Reply:
>What do you think about displaying the warning messages in the [Podman command handler](https://github.com/redhat-developer/odo/blob/main/pkg/dev/podmandev/command.go) instead, while the handler is walking through each component when running the command? You would still need to handle the case of K8s components not referenced by any command.

:facepalm:
I can't believe I didn't think of that.